lgc/state/PalMetadata.cpp

```diff
@@ -969,3 +920,179 @@ void PalMetadata::eraseFragmentInputInfo() {
+  static const std::map<unsigned, unsigned> shaderTable = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                           {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_HS_0},
+                                                           {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                           {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_GS_0},
+                                                           {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+  static const std::map<unsigned, unsigned> shaderTableGfx9 = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+
+  bool isGfx9 = m_pipelineState->getTargetInfo().getGfxIpVersion().major == 9;
+  const std::map<unsigned, unsigned> &currentShaderTable = (isGfx9 ? shaderTableGfx9 : shaderTable);
+  return currentShaderTable.at(callingConv);
+}
+
+// =====================================================================================================================
+// Returns to number of sgprs before the sgprs used for other purposes before the first user data sgps.
+//
+// @param conv : The calling convention of the shader
+unsigned PalMetadata::getNumberOfSgprsBeforeUserData(unsigned int conv) {
+  // GFX9+ merged shader have an extra 8 SGPRs before user data.
+  if (conv != CallingConv::AMDGPU_VS && m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9)
+    return 8;
+  return 0;
+}
+
+// =====================================================================================================================
+// Returns the offset of the userDataMapping from firstUserDataNode.  Returns UINT_MAX if it cannot be found.
+//
+// @param firstUserDataNode : An iterator identifing the starting point for the search.  It is assumed that it points to
+//                            the first user data node for some shader stage.
+// @param userDataMapping : The user data mapping that is being search for.
+//
+unsigned int
+PalMetadata::getOffsetOfUserDataReg(std::map<msgpack::DocNode, msgpack::DocNode>::iterator firstUserDataNode,
+                                    UserDataMapping userDataMapping) {
+
+  unsigned firstReg = firstUserDataNode->first.getUInt();
+  unsigned lastReg = firstReg + m_pipelineState->getTargetInfo().getGpuProperty().maxUserDataCount;
+  for (auto userDataNode = firstUserDataNode; userDataNode != m_registers.end(); ++userDataNode) {
```

Comment:
nit: This could written as:
```
for (auto& userDataNode : llvm::make_range(firstUderDataNode, m_registers.end()) {
  unsigned reg = userdataNode.first.getUInt();
  ...
```

Reply:
done